staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -479,3 +490,107 @@ func asPartialObjectMetadataList(result runtime.Object, groupVersion schema.Grou
 		return nil, newNotAcceptableError(fmt.Sprintf("no PartialObjectMetadataList exists in group version %s", groupVersion))
 	}
 }
+
+// watchListTransformerFunction an optional function
+// applied to watchlist bookmark events that transforms
+// the embedded object before sending it to a client.
+type watchListTransformerFunction func(watch.Event) watch.Event
+
+// createWatchListTransformerIfRequested returns a transformer function for watchlist bookmark event.
+//
+// note that: for simplicity, nil initialEventsListBlueprint indicates a no-op transformer.
+// otherwise, we assume a watchlist request for which we return a transformer.
+func createWatchListTransformerIfRequested(initialEventsListBlueprint runtime.Object, targetGVK *schema.GroupVersionKind, negotiatedEncoder runtime.Encoder) watchListTransformerFunction {
+	if initialEventsListBlueprint == nil {
+		return func(e watch.Event) watch.Event {
+			return e
+		}
+	}
+
+	return newWatchListTransformer(initialEventsListBlueprint, targetGVK, negotiatedEncoder).transform
+}
+
+// watchListTransformer performs transformation of
+// a special watchList bookmark event.
+//
+// The bookmark is annotated with InitialEventsListBlueprintAnnotationKey
+// and contains an empty, versioned list that we must encode in the requested format
+// (e.g., protobuf, JSON, CBOR) and then store as a base64-encoded string.
+type watchListTransformer struct {
+	initialEventsListBlueprint runtime.Object
+	targetGVK                  *schema.GroupVersionKind
+
+	negotiatedEncoder runtime.Encoder
+
+	buffer runtime.Splice
+}
+
+func newWatchListTransformer(initialEventsListBlueprint runtime.Object, targetGVK *schema.GroupVersionKind, negotiatedEncoder runtime.Encoder) *watchListTransformer {
+	return &watchListTransformer{
+		initialEventsListBlueprint: initialEventsListBlueprint,
+		targetGVK:                  targetGVK,
+		negotiatedEncoder:          negotiatedEncoder,
+		buffer:                     runtime.NewSpliceBuffer(),
+	}
+}
+
+func (e *watchListTransformer) transform(event watch.Event) watch.Event {
+	hasAnnotation, err := storage.HasInitialEventsEndBookmarkAnnotation(event.Object)
+	if err != nil {
+		return newWatchEventErrorFor(err)
+	}
+	if !hasAnnotation {
+		return event
+	}
+
+	encodedInitialEventsListBlueprint, err := e.encodeInitialEventsListBlueprint()
+	if err != nil {
+		return newWatchEventErrorFor(err)
+	}
+
+	// event.Object comes from a shared cache
+	// before modifying it we need to make a copy
```

Comment:
watch-list transformer will be called only for bookmarks, which are not coming from the shared cache.
So deep-copying it isn't needed, right?

Reply:
ah, good point, you are right, thanks.